Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -47,10 +51,111 @@ attribute [local instance] NumberField.ringOfIntegersAlgebra
 
 variable {K}
 
-theorem isUnit_iff_norm [NumberField K] (x : 𝓞 K) :
+theorem isUnit_iff_norm [NumberField K] {x : 𝓞 K} :
     IsUnit x ↔ |(RingOfIntegers.norm ℚ x : ℚ)| = 1 := by
   convert (RingOfIntegers.isUnit_norm ℚ (F := K)).symm
   rw [← abs_one, abs_eq_abs, ← Rat.RingOfIntegers.isUnit_iff]
 #align is_unit_iff_norm isUnit_iff_norm
 
 end IsUnit
+
+namespace NumberField.Units
+
+section coe
+
+/-- The `MonoidHom` from the group of units `(𝓞 K)ˣ` to the field `K`. -/
+def coe_to_field : (𝓞 K)ˣ →* K := (Units.coeHom K).comp (map (algebraMap (𝓞 K) K))
+
+variable {K}
+
+/-- The coercion of `x : (𝓞 K)ˣ` into `K`. -/
+@[coe] def to_field (x : (𝓞 K)ˣ) : K := coe_to_field K x
+
+variable (K)
+
+theorem coe_to_field_injective : Function.Injective (coe_to_field K) :=
+  fun _ _ h => Units.eq_iff.mp (SetCoe.ext h)
+
+/-- There is a natural coercion from `(𝓞 K)ˣ` to `(𝓞 K)` and then from `(𝓞 K)` to `K` but it is
+useful to also have a direct one from `(𝓞 K)ˣ` to `K`. -/
+instance : Coe (𝓞 K)ˣ K := ⟨to_field⟩
+
+@[ext]
+theorem ext {x y : (𝓞 K)ˣ} (h : (x : K) = y) : x = y := (coe_to_field_injective K).eq_iff.mp h
+
+@[simp]
+theorem map_pow (x : (𝓞 K)ˣ) (n : ℕ) : (x ^ n : K) = (x : K) ^ n :=
+  _root_.map_pow (coe_to_field K) x n
+
+@[simp]
+theorem map_one : ((1 : (𝓞 K)ˣ) : K) = 1 := rfl
+
+@[simp]
+theorem ne_zero (x : (𝓞 K)ˣ) : (x : K) ≠ 0 :=
+  Subtype.coe_injective.ne_iff.mpr (_root_.Units.ne_zero x)
+
+end coe
+
+open NumberField.InfinitePlace
+
+section torsion
+
+/-- The torsion subgroup of the group of units. -/
+def torsion : Subgroup (𝓞 K)ˣ := CommGroup.torsion (𝓞 K)ˣ
+
+theorem mem_torsion {x : (𝓞 K)ˣ} [NumberField K] :
+    x ∈ torsion K ↔ ∀ w : InfinitePlace K, w x = 1 := by
+  rw [eq_iff_eq (x : K) 1, torsion, CommGroup.mem_torsion, isOfFinOrder_iff_pow_eq_one]
+  refine ⟨fun ⟨n, h_pos, h_eq⟩ φ => ?_, fun h => ?_⟩
+  · refine norm_map_one_of_pow_eq_one φ.toMonoidHom (k := ⟨n, h_pos⟩) ?_
+    rw [PNat.mk_coe, ← map_pow, h_eq, map_one]
+  · obtain ⟨n, hn, hx⟩ := Embeddings.pow_eq_one_of_norm_eq_one K ℂ x.val.prop h
+    exact ⟨n, hn, by ext; rwa [map_pow, map_one]⟩
+
+instance : Nonempty (torsion K) := ⟨1⟩
```

Comment:
Is this really needed? It is a subgroup, so it should automatically inhabited.

Reply:
That's what I thought but without it, in the definition 
```lean
def torsion_order [NumberField K] : ℕ+ := ⟨Fintype.card (torsion K), Fintype.card_pos⟩
```
it times out while trying to infer `Nonempty { x // x ∈ torsion K }` for `Fintype.card_pos`.